File: src/python/challenges/problems/alice_and_bob_playing_flower_game_test.py

# https://leetcode.com/problems/alice-and-bob-playing-flower-game

from typing import List
from core.problem_base import *

class AliceAndBobPlayingFlowerGame(ProblemBase):
    def Solution(self, n: int, m: int) -> int:
        n_even = n // 2
        n_odd = n_even if n % 2 == 0 else n_even + 1

        m_even = m // 2
        m_odd = m_even if m % 2 == 0 else m_even + 1

        return n_even * m_odd + n_odd * m_even

if __name__ == '__main__':
    TestGen(AliceAndBobPlayingFlowerGame) \
        .Add(lambda tc: tc.Param(10).Param(3).Result(15) ) \
        .Add(lambda tc: tc.Param(1).Param(1).Result(0) ) \
        .Add(lambda tc: tc.Param(3).Param(2).Result(3) ) \
        .Run()

File: src/go/challenges/problems/alice_and_bob_playing_flower_game_test.go
// https://leetcode.com/problems/alice-and-bob-playing-flower-game

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type AliceAndBobPlayingFlowerGame struct{}

func TestAliceAndBobPlayingFlowerGame(t *testing.T) {
	gen := core.TestSuite[AliceAndBobPlayingFlowerGame]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(100000).Param(100000).Result(5000000000)
	}).Add(func(tc *core.TestCase) {
		tc.Param(10).Param(3).Result(15)
	}).Add(func(tc *core.TestCase) {
		tc.Param(1).Param(1).Result(0)
	}).Run(t)
}

func (AliceAndBobPlayingFlowerGame) Solution(n int, m int) int64 {
	nEven := int64(n / 2)
	nOdd := int64((n + 1) / 2)

	mEven := int64(m / 2)
	mOdd := int64((m + 1) / 2)

	return nEven*mOdd + nOdd*mEven
}
